xcode14.3 ios deploy target 11.0 compile error

halx99 opened the pull request:
Latest code can't compile ios when deployment target < 12.0, due to `std::get` for `std::variant` not supported, previous release works well.

If you don't want support ios deployment target < 12.0, just ignore this patch

kcat commented:
I'm confused. You say deployment target 11 fails to build, but the build seems to succeed with defining the macros for deployment target 11. `std::variant` is necessary with newer commits in either case.

halx99 commented:
Yes, maybe xcode version related, will check later

halx99 commented:
my local xcode version is 14.3, compile failed:
```txt
CompileC /Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/Objects-normal/arm64/voice.o /Users/bel/dev/openal-soft/core/voice.cpp normal arm64 c++ com.apple.compilers.llvm.clang.1_0.compiler (in target 'OpenAL' from project 'OpenAL')
    cd /Users/bel/dev/openal-soft
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -ivfsstatcache /var/folders/5p/_cbcxxp934l_cg5myf6d1ls80000gn/C/com.apple.DeveloperTools/14.3-14E222b/Xcode/SDKStatCaches.noindex/iphoneos16.4-20E238-.sdkstatcache -target arm64-apple-ios11.0 -fmessage-length\=167 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit\=0 -fcolor-diagnostics -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-float-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -Wno-implicit-fallthrough -DCMAKE_INTDIR\=\"Debug-iphoneos\" -DOpenAL_EXPORTS -DAL_BUILD_LIBRARY -DAL_ALEXT_PROTOTYPES -DALC_API\=__attribute__\(\(visibility\(\"default\"\)\)\) -DAL_API\=__attribute__\(\(visibility\(\"default\"\)\)\) -D__STDC_FORMAT_MACROS -DRESTRICT\=__restrict -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -g -Wno-sign-conversion -Wno-infinite-recursion -Wno-move -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -Wno-range-loop-analysis -Wno-semicolon-before-method-body -I/Users/bel/dev/openal-soft/build/Debug-iphoneos/include -I/Users/bel/dev/openal-soft/include -I/Users/bel/dev/openal-soft/build -I/Users/bel/dev/openal-soft -I/Users/bel/dev/openal-soft/common -I/Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/DerivedSources-normal/arm64 -I/Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/DerivedSources/arm64 -I/Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/DerivedSources -F/Users/bel/dev/openal-soft/build/Debug-iphoneos -F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/System/Library/Frameworks -std\=gnu++17 -fPIC -fvisibility\=hidden -Winline -Wunused -Wall -Wextra -Wshadow -Wconversion -Wcast-align -Wpedantic -Wold-style-cast -Wnon-virtual-dtor -Woverloaded-virtual -Wno-c++20-attribute-extensions -fno-math-errno -pthread -MMD -MT dependencies -MF /Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/Objects-normal/arm64/voice.d --serialize-diagnostics /Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/Objects-normal/arm64/voice.dia -c /Users/bel/dev/openal-soft/core/voice.cpp -o /Users/bel/dev/openal-soft/build/build/OpenAL.build/Debug-iphoneos/Objects-normal/arm64/voice.o
In file included from /Users/bel/dev/openal-soft/core/voice.cpp:24:
/Users/bel/dev/openal-soft/core/async_event.h:60:17: error: 'get<AsyncSourceStateEvent, std::monostate, AsyncSourceStateEvent,
      AsyncBufferCompleteEvent, AsyncEffectReleaseEvent, AsyncDisconnectEvent>' is unavailable: introduced in iOS 12.0
    return std::get<T>(*evt);
                ^
/Users/bel/dev/openal-soft/core/voice.cpp:229:17: note: in instantiation of function template specialization
      'InitAsyncEvent<AsyncSourceStateEvent>' requested here
    auto &evt = InitAsyncEvent<AsyncSourceStateEvent>(evt_vec.first.buf);
                ^
In file included from /Users/bel/dev/openal-soft/core/voice.cpp:4:
In file included from /Users/bel/dev/openal-soft/core/voice.h:4:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/array:127:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/algorithm:1712:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/memory:877:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/iterator:684:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/__iterator/common_iterator.h:22:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/variant:1539:16: note: 
      'get<AsyncSourceStateEvent, std::monostate, AsyncSourceStateEvent, AsyncBufferCompleteEvent, AsyncEffectReleaseEvent, AsyncDisconnectEvent>' has been explicitly
      marked unavailable here
constexpr _Tp& get(variant<_Types...>& __v) {
               ^
In file included from /Users/bel/dev/openal-soft/core/voice.cpp:24:
/Users/bel/dev/openal-soft/core/async_event.h:60:17: error: 'get<AsyncBufferCompleteEvent, std::monostate, AsyncSourceStateEvent,
      AsyncBufferCompleteEvent, AsyncEffectReleaseEvent, AsyncDisconnectEvent>' is unavailable: introduced in iOS 12.0
    return std::get<T>(*evt);
                ^
/Users/bel/dev/openal-soft/core/voice.cpp:1153:25: note: in instantiation of function template specialization
      'InitAsyncEvent<AsyncBufferCompleteEvent>' requested here
            auto &evt = InitAsyncEvent<AsyncBufferCompleteEvent>(evt_vec.first.buf);
                        ^
In file included from /Users/bel/dev/openal-soft/core/voice.cpp:4:
In file included from /Users/bel/dev/openal-soft/core/voice.h:4:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/array:127:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/algorithm:1712:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/memory:877:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/iterator:684:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/__iterator/common_iterator.h:22:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/variant:1539:16: note: 
      'get<AsyncBufferCompleteEvent, std::monostate, AsyncSourceStateEvent, AsyncBufferCompleteEvent, AsyncEffectReleaseEvent, AsyncDisconnectEvent>' has been
      explicitly marked unavailable here
constexpr _Tp& get(variant<_Types...>& __v) {
               ^
2 errors generated.
```

cmake configure output:
```txt
-- The C compiler identification is AppleClang 14.0.3.14030022
-- The CXX compiler identification is AppleClang 14.0.3.14030022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ - skipped
-- Detecting CXX compile features
```

halx99 commented:
Confirmed, xcode14.3 not working, tested xcode13~xcode14.2 works well
